lib/Optimizer/Transforms/StatePreparation.cpp

```diff
@@ -0,0 +1,209 @@
+  assert(attr.has_value());
+
+  auto elementsAttr = dyn_cast<mlir::ElementsAttr>(attr.value());
+  assert(elementsAttr);
+  auto eleTy = elementsAttr.getElementType();
+  auto values = elementsAttr.getValues<mlir::Attribute>();
+
+  for (auto it = values.begin(); it != values.end(); ++it) {
+    auto valAttr = *it;
+
+    auto v = [&]() -> std::complex<double> {
+      if (isa<FloatType>(eleTy))
+        return {cast<FloatAttr>(valAttr).getValue().convertToDouble(),
+                static_cast<double>(0.0)};
+      if (isa<IntegerType>(eleTy))
+        return {static_cast<double>(cast<IntegerAttr>(valAttr).getInt()),
+                static_cast<double>(0.0)};
+      assert(isa<ComplexType>(eleTy));
+      auto arrayAttr = cast<mlir::ArrayAttr>(valAttr);
+      auto real = cast<FloatAttr>(arrayAttr[0]).getValue().convertToDouble();
+      auto imag = cast<FloatAttr>(arrayAttr[1]).getValue().convertToDouble();
+      return {real, imag};
+    }();
+
+    result.push_back(v);
+  }
+  return result;
+}
+
+LogicalResult transform(ModuleOp module, func::FuncOp funcOp) {
+  auto builder = OpBuilder::atBlockBegin(&funcOp.getBody().front());
+  auto toErase = std::vector<mlir::Operation *>();
+  auto hasInitState = false;
+  auto replacedInitState = false;
+
+  funcOp->walk([&](Operation *op) {
+    if (auto initOp = dyn_cast<quake::InitializeStateOp>(op)) {
+      toErase.push_back(initOp);
+      hasInitState = true;
+      auto loc = op->getLoc();
+      builder.setInsertionPointAfter(initOp);
+      // Find the qvector alloc.
+      auto qubits = initOp.getOperand(0);
+      if (auto alloc = dyn_cast<quake::AllocaOp>(qubits.getDefiningOp())) {
+
+        // Find vector data.
+        auto data = initOp.getOperand(1);
+        if (auto cast = dyn_cast<cudaq::cc::CastOp>(data.getDefiningOp())) {
+          data = cast.getOperand();
+          toErase.push_back(cast);
+        }
+        if (auto addr =
+                dyn_cast<cudaq::cc::AddressOfOp>(data.getDefiningOp())) {
+
+          auto globalName = addr.getGlobalName();
+          auto symbol = module.lookupSymbol(globalName);
+          if (auto global = dyn_cast<cudaq::cc::GlobalOp>(symbol)) {
+            // Read state initialization data from the global array.
+            auto vec = readConstantArray(builder, global);
+
+            // Prepare state from vector data.
+            auto gateBuilder = StateGateBuilder(builder, loc, qubits);
+            auto decomposer = StateDecomposer(gateBuilder, vec);
+            decomposer.decompose();
+
+            initOp.replaceAllUsesWith(qubits);
+            toErase.push_back(addr);
+            toErase.push_back(global);
+            replacedInitState = true;
+          }
+        }
+      }
+    }
+  });
+
+  if (hasInitState && !replacedInitState) {
+    funcOp.emitOpError("StatePreparation failed to replace quake.init_state");
+    return failure();
+  }
+
+  for (auto &op : toErase) {
+    if (op->getUses().empty()) {
+      op->erase();
```

Comment:
It isn't necessary for a pass to keep track of and delete ops that are unused in general. It's less tedious and clearer code to just let region simplification get rid of the dead code.

That said, an operation such as cc::StoreOp has side-effects and as such region simplification will not erase those.

Reply:
i think it does not erase the `quake.init_state` either (it remained and failed to lower to llvm in my tests), and this is the main purpose here (plus all its dependencies)

Reply:
Had to keep it - if I don't remove any of them we either end up with invalid IR or quantum passes fail